File: src/main/java/akostenko/aicars/menu/Mode.java
package akostenko.aicars.menu;

import java.util.Objects;

public abstract class Mode implements MenuItem {
    public static Mode forName(String name) {
        switch (name) {
            case WithPlayer.NAME: return new WithPlayer();
            case CarPerformanceTests.NAME: return new CarPerformanceTests();
            case CarPhysicsTests.NAME: return new CarPhysicsTests();
            default: return defaultMode();
        }
    }

    private static Mode defaultMode() {
        return new WithPlayer();
    }

    @Override
    public int hashCode() {
        return Objects.hash(getTitle());
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof Mode && Objects.equals(getTitle(), ((Mode) obj).getTitle());
    }
}

File: src/main/java/akostenko/aicars/menu/Mode.kt
package akostenko.aicars.menu

import java.util.Objects

abstract class Mode : MenuItem {

    override fun hashCode(): Int {
        return Objects.hash(title)
    }

    override fun equals(other: Any?): Boolean {
        return other is Mode && title == other.title
    }

    companion object {
        fun forName(name: String): Mode {
            when (name) {
                WithPlayer.NAME -> return WithPlayer()
                CarPerformanceTests.NAME -> return CarPerformanceTests()
                CarPhysicsTests.NAME -> return CarPhysicsTests()
                else -> return defaultMode()
            }
        }

        private fun defaultMode(): Mode = WithPlayer()
    }
}